### ISSUE TITLE
Inject slice of interfaces

### ISSUE THREAD

**Issue by alexmatsak** [google/wire#148]
Assume that i have this code
```
type Counter interface {
	GetCount() int
}

type A struct{}
type B struct{}

func (*A) GetCount() int {
	return 1
}

func (*B) GetCount() int {
	return 2
}

type ChainCounter struct {
	counters []Counter
}

// Providers

func ProvideA() *A {
	return &A{}
}

func ProvideB() *A {
	return &A{}
}

func ProvideChainCounter(counters ...Counter) *ChainCounter {
	return &ChainCounter{counters: counters}
}
```

How can i inject `A` and `B` (which implementing `Counter`) to chain counter with wire ?

**Comment by zombiezen** [comment]
(I'm assuming `ProvideB` was supposed to return `*B`.)

```go
func inject() *ChainCounter {
  wire.Build(
    ProvideA,
    ProvideB,
    ProvideCounterList,
    ProvideChainCounter,
  )
  return nil
}

func ProvideCounterList(a *A, b *B) []Counter {
  return []Counter{a, b}
}
```

**Comment by shantuo** [comment]
Actually you only need a `ProvideCounterList`, not the `ProvideChainCounter`.

```go
func inject() *ChainCounter {
  wire.Build(
    ProvideA,
    ProvideB,
    ProvideCounterList,
    ChainCounter{}, // This will be a wire.Struct(new(ChainCounter), "*") after #147 is merged.
  )
  return nil
}

func ProvideCounterList(a *A, b *B) []Counter {
  return []Counter{a, b}
}
```

**Comment by alexmatsak** [comment]
thanks!

**Comment by coolyrat** [comment]
I'm facing a similar question. I'm using gin web framework, and there's a func type call gin.HandlerFunc .  It would be nice to have wire group the types, is there a way to achieve?


```go
var serverSet = wire.NewSet(
	...
	provideLogMiddleware,
	provideErrorHandlerMiddleware,
	provideMiddlewares,
)

func provideErrorHandlerMiddleware(logger *zap.Logger) gin.HandlerFunc {
	return middleware.ErrorHandler(logger)
}

func provideLogMiddleware() gin.HandlerFunc {
	return gin.Logger()
}

func provideMiddlewares(middleware ...gin.HandlerFunc) []gin.HandlerFunc {
	return middleware
}
```

**Comment by vangent** [comment]
You need to explicitly group the types yourself; there is no way for `wire` to know which `gin.HandlerFunc`s to include otherwise.

So you can either do:

```
func provideErrorHandlerMiddleware(logger *zap.Logger) *middleware.ErrorHandler {
	return middleware.ErrorHandler(logger)
}
func provideLogMiddleware() *gin.Logger {
	return gin.Logger()
}
...
func provideMiddlewares(a *middleware.ErrorHandler, b *gin.Logger) []gin.HandlerFunc {
	return []gin.Handlerfunc{a, b}
}
```

(this is similar to the original example given [here](https://github.com/google/wire/issues/148#issuecomment-474964848); you may need to add a `wire.Struct` provider as described there).

OR

```
func provideMiddlewares(logger *zap.Logger) []gin.HandlerFunc {
	return []gin.Handlerfunc{middleware.ErrorHandler(logger), gin.Logger() }
}
```

The first option is probably better for `wire`, since `wire` can help construct all of the `gin.HandlerFunc` implementations.

HTH

**Comment by coolyrat** [comment]
thanks for making this clear! I will go for the first option =D

**Comment by alexmatsak** [comment]
Is it possible to inject **nil** to  interface param? 

For example
```
type A struct {
    r io.Reader
}

func NewA(r io.Reader) *A {
    return &A{r:r}
}

```

Try to use `wire.InterfaceValue(new(io.Reader), nil)` but got error that _nil_ doesn`t implement _Reader_ interface

**Comment by vangent** [comment]
Try `wire.InterfaceValue(new(io.Reader), io.Reader(nil))`.
